src/Compilers/CSharp/Portable/Symbols/NamespaceSymbol.cs

```diff
@@ -311,6 +311,23 @@ internal NamespaceSymbol GetNestedNamespace(NameSyntax name)
             return null;
         }
 
+        private ImmutableArray<NamedTypeSymbol> _lazyTypesWithExtensionMethods;
+        private ImmutableArray<NamedTypeSymbol> TypesWithExtensionMethods
+        {
+            get
+            {
+                var typesWithExtensionMethods = this._lazyTypesWithExtensionMethods;
+                if (typesWithExtensionMethods.IsDefault)
+                {
+                    this._lazyTypesWithExtensionMethods = this.GetTypeMembersUnordered().WhereAsArray(t => t.MightContainExtensionMethods);
```

Comment:
We do not need barrier here. All platforms we run on honor data dependency. - You cannot read the referent before the reference.

A while ago we have decided to code against CLR2.0 memory model and that is all that we can test anyways. We have hundreds (or maybe thousands) places with code relying on data dependency directly or indirectly. 

When we run on Alphas, we may need a serious audit :-)